### DISCUSSION TITLE
Logtalk and CHR

### DISCUSSION THREAD

**Discussion by vsraptor** [LogtalkDotOrg/logtalk3#88] [Q&A]
Is there any tutorials on integrating CHR with LogTalk ?

Questions like : 

Can you have separate constraint stores i.e. one per object ? Or is it everything global ?
If scoped how do you access them ?
How do you use them ...etc.. f.e.

obj1::fun(abc) <=> obj2::funx(blah) 

What language patterns/nomenclatures are good practice ? How'd encapsulation,inheritance work ?
prototypes ? 

Swi-prolog here.

**Comment by pmoura** [comment]
The CHR library distributed with SWI-Prolog requires CHR code to be in a Prolog module. There's no workaround available that would work with CHR code in a Logtalk object. The root issue is that this CHR implementation makes calls to the `prolog_load_context/2` predicate to get the module name, which doesn't exist when using a Logtalk object. This CHR library can still be used in Logtalk applications but you will need to keep the CHR code inside modules. Those modules can send messages to Logtalk objects and Logtalk objects can call those module predicates.

If your use of CHR is light, the Logtalk distribution includes (in the [`ports/toychr`](https://github.com/LogtalkDotOrg/logtalk3/tree/master/ports/toychr) directory) a port of Gregory J. Duck's ToyCHR implementation with some [examples](https://github.com/LogtalkDotOrg/logtalk3/tree/master/ports/toychr/examples). But the changes to the original ToyCHR in this port were the minimal ones to allow the examples of CHR rules encapsulated inside Logtalk objects to run.

**Comment by vsraptor** [comment]
from the examples toychr seems to supports the full CHR syntax ?
What things it does not support !?
Or is it simply slower or something ... thanks

**Reply by pmoura** [reply]
I would expect it to be slower. But have not tried to compare it with other CHR implementations. The original website page for ToyCHR is no longer available but the author pages are still up at:

https://www.comp.nus.edu.sg/~gregory/

I suggest you contact the original author to find more about any ToyCHR limitations. You can also try your own examples, or port examples from e.g. SWI-Prolog and check if you get them to run and giving the same results (if you do, your contributions are welcome).

P.S. I committed two other classical examples: Fibonacci sequence and prime numbers.

**Reply by vsraptor** [reply]
>> This CHR library can still be used in Logtalk applications but you will need to keep the CHR code inside modules.

So in a sense you have one Global CHR constraint store..?
I can see how you can call Logtalk from CHR...i.e. like prolog but with obj::blah() 
but how do you call from LT object a CHR rule  ?

**Reply by pmoura** [reply]
When sending messages to Logtalk objects from CHR code, be sure to compile and run the Logtalk code with the  `optimize` flag turned on. Loading the Logtalk code before the CHR code will ensure that the `::/2` goals will be optimized, thus removing (assuming both arguments bound) any message sending overhead.

To call from a Logtalk object (or category) a CHR predicate, either use explicit module-qualification (using the `:/2` operator) or add a [`use_module/2`](https://logtalk.org/manuals/refman/directives/use_module_2.html) directive in the object or category (be sure to read the documentation of the directive, notably on how it differs from the Prolog own `use_module/2` directive).
